File: cs50/credit/credit.c
#include <stdio.h>
#include <cs50.h>

bool is_valid(unsigned long long card_number);
string card_type(unsigned long long card_number);


int main(void)
{
    unsigned long long card_number;

    do
    {
        card_number = get_long("Number: ");
    }
    while (card_number <= 0);

    if (is_valid(card_number))
    {
        printf("%s\n", card_type(card_number));
    }
    else
    {
        printf("INVALID\n");
    }

    return 0;
}


bool is_valid(unsigned long long card_number)
{
    int sum = 0;
    bool multiply = false;

    while (card_number > 0)
    {
        int digit = card_number % 10;
        card_number /= 10;

        if (multiply)
        {
            digit *= 2;
            digit = (digit % 10) + (digit / 10);
        }

        sum += digit;
        multiply = !multiply;
    }

    return (sum % 10 == 0);
}

string card_type(unsigned long long card_number)
{
    int length = 0;
    unsigned long long temp = card_number;

    while (temp > 0)
    {
        temp /= 10;
        length++;
    }

    if ((length == 15) && ((card_number / 10000000000000 == 34) || (card_number / 10000000000000 == 37)))
    {
        return "AMEX";
    }
    else if ((length == 16) && ((card_number / 100000000000000 == 51) || (card_number / 100000000000000 == 52) || (card_number / 100000000000000 == 53) || (card_number / 100000000000000 == 54) || (card_number / 100000000000000 == 55)))
    {
        return "MASTERCARD";
    }
    else if (((length == 13) || (length == 16)) && ((card_number / 1000000000000000 == 4) || (card_number / 1000000000000 == 4))) // Corrected condition for VISA
    {
        return "VISA";
    }
    else
    {
        return "INVALID";
    }
}

File: cs50/sentimental-credit/credit.py
def get_long(prompt):
    while True:
        try:
            return int(input(prompt))
        except ValueError:
            print("Invalid input. Please enter a long integer.")

def is_valid(card_number):
    sum = 0
    multiply = False

    while card_number > 0:
        digit = card_number % 10
        card_number //= 10

        if multiply:
            digit *= 2
            digit = (digit % 10) + (digit // 10)

        sum += digit
        multiply = not multiply

    return sum % 10 == 0

def card_type(card_number):
    length = 0
    temp = card_number

    while temp > 0:
        temp //= 10
        length += 1

    if (length == 15) and (card_number // 10000000000000 == 34 or card_number // 10000000000000 == 37):
        return "AMEX"
    elif (length == 16) and (51 <= card_number // 100000000000000 <= 55):
        return "MASTERCARD"
    elif (length == 13 or length == 16) and (card_number // 1000000000000 == 4 or card_number // 1000000000000000 == 4):
        return "VISA"
    else:
        return "INVALID"

def main():
    card_number = 0

    while card_number <= 0:
        card_number = get_long("Number: ")

    if is_valid(card_number):
        print(card_type(card_number))
    else:
        print("INVALID")


main()
